### PR TITLE
Series of Changes to Help Nikki Build

### PR DESCRIPTION
**Pull request by changlinli**
These changes allow Nikki to build for GHC versions 7.8.x and 7.6.x, as verified by Travis. The step-by-step instructions for how to do this on Linux are also laid out in the src README.

### CODE REVIEW

**Comment by changlinli** [comment]
Travis somehow barfed on GHC 7.8, when I added a GHC 7.4 option, which makes no sense. I think GitHub somehow managed to mess up the Travis build, by causing it to fail at [https://travis-ci.org/nikki-and-the-robots/nikki/builds/81238330][1] and simultaneously hang at [https://travis-ci.org/changlinli/nikki/jobs/81238328][2] (which now passes because I forced it to rerun), but I don't know how that would happen...

[1]: https://travis-ci.org/nikki-and-the-robots/nikki/builds/81238330
[2]: https://travis-ci.org/changlinli/nikki/jobs/81238328

**Comment by soenkehahn** [comment]
This is awesome!

I've restarted the failing 7.8 build and now it passes. My guess is that we have some non-deterministic build failures. That's not good and maybe we can fix that in the future. But this is still awesome!!! :)

@changlinli: I would have some comments on your changes, but this is already much better than before. Should I merge or do you want to wait for a few days until I find the time to write some comments?

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  - 7.4
+
+before_install:
+  - sudo apt-get update
+  - sudo apt-get install haskell-platform g++ cmake pkg-config libzip-dev libopenal-dev libsndfile1-dev
+
+install:
+  - cd src
+  - ./linuxCompile.sh --extra-include-dir=/usr/include/AL --constraint=bindings-libzip==0.10.*
+  - cabal install -j --only-dependencies --constraint=bindings-libzip==0.10.* --extra-include-dir=/usr/include/AL && cabal configure --enable-tests && cabal build && cabal test
```
Comment (soenkehahn) [inline .travis.yml]: `linuxCompile` already calls `cabal` so this builds `nikki` twice. Maybe we can just call `cabal test` after `linuxCompile`. (We have to pass `--enable-tests` to it then, I guess.)

File: src/Distribution/AutoUpdate.hs [hunk 2]
```diff
@@ -53,6 +53,18 @@ import Distribution.AutoUpdate.VerifySignatures
...
+#if MIN_VERSION_base(4,7,0)
+#else
+-- Taken directly from the source of base version 4.8.0.0
+instance Foldable (Either a) where
+    foldMap _ (Left _) = mempty
+    foldMap f (Right y) = f y
+
+    foldr _ z (Left _) = z
+    foldr f z (Right y) = f y z
+#endif
```
Comment (soenkehahn) [inline src/Distribution/AutoUpdate.hs]: There's a package called `base-compat` that tries to provide a compatibility layer on top of different versions of `base`. Maybe we could use that. We would get rid of `CPP` then.

Comment (soenkehahn) [inline src/Distribution/AutoUpdate.hs]: (Same for most of the other uses of `CPP` in this PR.)

**Comment by changlinli** [comment]
For what it's worth, I suspect it's a GitHub/Travis issue rather than a build issue that's causing the non-determinism. Somehow your Travis build died on 7.8.x at the exact same time mine did and while yours errored out, mine just hung. Maybe Travis is doing some sort of caching/sharing of identical builds?

I'm happy to edit my PR. Didn't know `base-compat` was a thing... learn something from open source stuff every day. If you have any other comments in the upcoming days let me know. Depending on my schedule though it may make sense for you to merge it in and apply those changes yourself if you don't hear from me for a couple of days. Weekdays can be a bit rough for me.

Now off to implement those changes.

**Comment by changlinli** [comment]
Somewhat surprisingly it does not look like it's possible to replace any of the CPP pragmas apart from the orphan instance of Foldable since base-compat does not export the function `tryReadMVar`. I don't know why that is, except it looks like GHC implements it with prim-ops magic. I have no experience with MVars so I have no idea whether that's semantically the same as what is implemented in Utils (although a cursory look seems to suggest so...).

**Comment by changlinli** [comment]
And even that doesn't work :(. `base-compat`'s `Prelude.Compat` doesn't provide the instance of `Foldable` for `Either`, that shows up in `base-orphans`, so there's going to be an orphan instance no matter what. In light of that, I've reverted my commit getting rid of the manual orphan instance of Foldable. 

Do you wanna pull in `base-orphans` or just go ahead as is with the manual instance?

**Comment by soenkehahn** [comment]
I would prefer to pull in `base-orphans`.

Anyways, I think I'm going to merge this soon. Thanks very much for your work!

**Comment by soenkehahn** [comment]
One more general comment: I think we should have a cabal freeze file. That way we could pin down the exact versions that work for `nikki`.

I don't use the haskell platform and a freeze file would allow me to reproduce your build exactly. Hopefully.

**Comment by changlinli** [comment]
Gah! The build on Travis is dying with an error in one of the dependencies inexplicably again for GHC 7.8.3 and once again it hangs on my Travis build at the exact same time (since each commit triggers both a build here and a build for me) and is fine after a restart.

**Comment by changlinli** [comment]
I've put in the `cabal.config` file that results from freezing my build and suffixed it with `.inactive` so that if users want a last known good configuration they have it, but that it isn't used by default. I've made it inert since Nikki builds fine against several different versions of GHC and an active freeze file would prevent users with certain GHC versions from building Nikki without first moving the cabal.config file.

**Pull request merged**